src/main/java/pl/edu/icm/pl/mxrdr/extension/workflow/XdsImagesPatternStep.java

```diff
@@ -0,0 +1,100 @@
+package pl.edu.icm.pl.mxrdr.extension.workflow;
+
+import edu.harvard.iq.dataverse.dataset.datasetversion.DatasetVersionServiceBean;
+import edu.harvard.iq.dataverse.workflow.WorkflowContext;
+import edu.harvard.iq.dataverse.workflow.step.WorkflowStepResult;
+import io.vavr.Tuple;
+import io.vavr.Tuple2;
+import org.apache.commons.lang3.StringUtils;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Computes file name pattern to pass to <code>generate_XDS.INP</code> script.
+ */
+class XdsImagesPatternStep extends XdsWorkflowStep {
+
+    static final String STEP_ID = "xds-images-pattern";
+
+    /**
+     * Input parameter containing {@value #FILE_NAMES_SEPARATOR} separated list of file names to compute pattern for.
+     */
+    static final String FILE_NAMES_PARAM_NAME = "fileNames";
+
+    /**
+     * Values separator for {@value #FILE_NAMES_PARAM_NAME} input parameter.
+     */
+    static final String FILE_NAMES_SEPARATOR = ";";
+
+    private final List<String> fileNames;
+
+    public XdsImagesPatternStep(DatasetVersionServiceBean datasetVersions, Map<String, String> inputParams) {
+        super(datasetVersions);
+        this.fileNames = asList(inputParams.getOrDefault(FILE_NAMES_PARAM_NAME, "").split(FILE_NAMES_SEPARATOR));
+    }
+
+    @Override
+    protected WorkflowStepResult runInternal(WorkflowContext context, Path workDir) {
+        String namePattern = calculatePattern();
+        // FIXME: save name pattern in result
+        return WorkflowStepResult.OK;
+    }
+
+    String calculatePattern() {
+        Tuple2<String, List<String>> prefixTuple = longestCommonPrefix(fileNames);
+        String prefix = prefixTuple._1;
+
+        Tuple2<String, List<String>> suffixTuple = longestCommonPrefix(
+                prefixTuple._2.stream().map(StringUtils::reverse).collect(toList()));
+        String suffix = StringUtils.reverse(suffixTuple._1);
+
+        if (".h5".equals(suffix)) {
+            return prefix + "master.h5";
+        } else {
+            int questionMarkCount = prefixTuple._2.get(0).length() - prefix.length() - suffix.length();
+            String questionMarks = new String(new char[questionMarkCount]).replace('\0', '?');
+            return prefix + questionMarks + suffix;
+        }
+    }
+
+    private Tuple2<String, List<String>> longestCommonPrefix(List<String> candidates) {
```

Comment:
Ad 1. Added extensive comment on the to of the class describing it.
Ad 2. Fixed and added a test case for that.
Ad 3. Memory gain would be quite insignificant, as it would only free up a few keys, that are common substrings, hence small and limited in number anyways; was thinking about it earlier, but didn't seem worth it.